### PR TITLE
INT-7586 - Implementing Users, devices, and application steps

### PR DESCRIPTION
**Pull request by adam-in-ict**
# Description

Thank you for contributing to a JupiterOne integration!

Please include a summary of the change and which issue is fixed. Please also
include relevant motivation and context. List any dependencies that are required
for this change.

## Type of change

Please leave any irrelevant options unchecked.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [ ] This change requires a documentation update

### General Development Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### Integration Development Checklist:

Please leave any irrelevant options unchecked.

- [ ] I have checked for additional permissions required to call any new API
      endpoints, and have documented any additional permissions in
      `jupiterone.md`, where necessary.
- [ ] My changes properly paginate the target service provider's API
- [ ] My changes properly handle rate limiting of the target service provider's
      API
- [ ] My new integration step is instrumented to execute in the correct order
      using `dependsOn`
- [ ] I have referred to the
      [JupiterOne data model](https://github.com/JupiterOne/data-model/tree/main/src/schemas)
      to ensure that any new entities/relationships, and relevant properties,
      match the recommended model for this class of data
- [ ] I have updated the `CHANGELOG.md` file to describe my changes
- [ ] When changes include modifications to existing graph data ingestion, I've
      reviewed all existing managed questions referencing the entities,
      relationships, and their property names, to ensure those questions still
      function with my changes.

### CODE REVIEW

File: src/steps/application/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import { Steps } from '../constants';
+
+// See test/README.md for details
+let recording: Recording;
+afterEach(async () => {
+  await recording.stop();
+});
+
+// We currently have to skip this until we get applications added to a test environment
+test.skip('fetch-applications', async () => {
```
Comment (adam-in-ict) [inline src/steps/application/index.test.ts:13]: We're currently unable to get our sandbox populated with test data for this step.

File: src/steps/applicationCategory/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import { Steps } from '../constants';
+
+// See test/README.md for details
+let recording: Recording;
+afterEach(async () => {
+  await recording.stop();
+});
+
+// We currently have to skip this until we get application categories added to a test environment
+test.skip('fetch-application-categories', async () => {
```
Comment (adam-in-ict) [inline src/steps/applicationCategory/index.test.ts:13]: We're currently unable to get our sandbox populated with test data for this step.

File: src/steps/application/__recordings__/fetch-applications_2367457233/recording.har [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+          "headersSize": 388,
+          "httpVersion": "HTTP/1.1",
+          "method": "POST",
+          "postData": {
+            "mimeType": "application/json",
+            "params": [],
+            "text": "[REDACTED]"
+          },
+          "queryString": [],
+          "url": "https://api.umbrella.com//auth/v2/token"
```
Comment (VDubber) [inline src/steps/application/__recordings__/fetch-applications_2367457233/recording.har:67]: This URL has a double //.

File: docs/spec/src/device/index.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { RelationshipClass, StepSpec } from '@jupiterone/integration-sdk-core';
+import { IntegrationConfig } from '../../../../src/config';
+
+export const deviceSpec: StepSpec<IntegrationConfig>[] = [
+  {
+    /**
+     * ENDPOINT: https://<baseurl>/deployments/v2/networkdevices
+     * PATTERN: Fetch Entities
+     */
+    id: 'fetch-users',
```
Comment (VDubber) [inline docs/spec/src/device/index.ts:10]: This file is for devices.

File: src/steps/constants.ts [hunk 5]
```diff
@@ -42,6 +62,11 @@ export const Entities: Record<
     _type: 'cisco_umbrella_destination_list',
     _class: ['Record'],
   },
+  DEVICE: {
+    resourceName: 'Network Device',
+    _type: 'cisco_umbrella_device',
```
Comment (VDubber) [inline src/steps/constants.ts:67]: rename _type: `cisco_umbrella_network_device`?

File: src/steps/role/converter.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import {
+  createIntegrationEntity,
+  Entity,
+} from '@jupiterone/integration-sdk-core';
+import { Role } from '../../types';
+
+import { Entities } from '../constants';
+
+export function createRoleKey(id: number) {
+  return `${Entities.ROLE._type}:${id}`;
```
Comment (VDubber) [inline src/steps/role/converter.ts:10]: I like this pattern, we should consider standardizing it in the SDK

Comment (VDubber) [reply src/steps/role/converter.ts:10]: Or we should make callisto give us real customer data :)

**Review by VDubber — APPROVED** [review]

File: src/steps/application/__recordings__/fetch-applications_2367457233/recording.har (hunk 3 above)
Comment (adam-in-ict) [reply src/steps/application/__recordings__/fetch-applications_2367457233/recording.har]: Yep.  Was due to having an extra / at the end of the base URL.  I've removed it and re-recorded all .har files in the latest commit.

File: docs/spec/src/device/index.ts (hunk 4 above)
Comment (adam-in-ict) [reply docs/spec/src/device/index.ts]: Whoops!  Updated in a recent commit.

File: src/steps/constants.ts (hunk 5 above)
Comment (adam-in-ict) [reply src/steps/constants.ts]: Renamed in a recent commit.

**Review by VDubber — APPROVED** [review]

**Pull request merged**